### PR TITLE
qt@5: qtwebengine 5.15.8

### PR DESCRIPTION
**Pull request by Bo98**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Security fixes and hopefully build fixes for Monterey too.

### CODE REVIEW

**Comment by Bo98** [comment]
#90046 will hopefully fix the `pyside@2` linkage.

I might roll this into the icu4c PR to avoid double revision bumps.

File: Formula/qt@5.rb [hunk 1]
```diff
@@ -179,8 +181,7 @@ def caveats
 
     if Hardware::CPU.arm?
       s += <<~EOS
-
```
Comment (Bo98) [inline Formula/qt@5.rb:182]: Didn't mean to remove this line - I'll add it back.

**Comment by danielnachun** [comment]
I think this was the same error we were dealing with before on Monterey (https://github.com/Homebrew/homebrew-core/issues/87708#issuecomment-960423003):
```
2021-11-27T20:55:29.0234890Z TOOL_VERSION=1638034840 ../../3rdparty/chromium/build/toolchain/mac/linker_driver.py -Wcrl,strippath,/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/strip ../../../../../../../../Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -B /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/  -Wcrl,dsym,. -Wcrl,dsymutilpath,../../../../../../../../Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/dsymutil  -Wl,-fatal_warnings -stdlib=libc++ -arch x86_64 -Wl,-no_data_in_code_info -Wl,-no_function_starts -isysroot ../../../../../../../../Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.0.sdk -mmacosx-version-min=10.13 -Wl,-ObjC -Wcrl,strip,-x,-S -o "./transport_security_state_generator" -Wl,-filelist,"./transport_security_state_generator.rsp" -framework ApplicationServices -framework AppKit -framework CoreFoundation -framework IOKit -framework OpenDirectory -framework Security -framework Foundation  -lbsm -lpmenergy -lpmsample
2021-11-27T20:55:29.0240990Z Undefined symbols for architecture x86_64:
2021-11-27T20:55:29.0242210Z   "std::__1::basic_string<unsigned short, base::string16_internals::string16_char_traits, std::__1::allocator<unsigned short> >::shrink_to_fit()", referenced from:
2021-11-27T20:55:29.0243690Z       base::UTF8ToUTF16(char const*, unsigned long, std::__1::basic_string<unsigned short, base::string16_internals::string16_char_traits, std::__1::allocator<unsigned short> >*) in libbase.a(base_jumbo_20.o)
2021-11-27T20:55:29.0245600Z       base::WideToUTF16(wchar_t const*, unsigned long, std::__1::basic_string<unsigned short, base::string16_internals::string16_char_traits, std::__1::allocator<unsigned short> >*) in libbase.a(base_jumbo_20.o)
2021-11-27T20:55:29.0246770Z ld: symbol(s) not found for architecture x86_64
2021-11-27T20:55:29.0248830Z clang: error: linker command failed with exit code 1 (use -v to see invocation)
2021-11-27T20:55:29.0249570Z Traceback (most recent call last):
2021-11-27T20:55:29.0250280Z   File "../../3rdparty/chromium/build/toolchain/mac/linker_driver.py", line 307, in <module>
2021-11-27T20:55:29.0250910Z     Main(sys.argv)
2021-11-27T20:55:29.0251530Z   File "../../3rdparty/chromium/build/toolchain/mac/linker_driver.py", line 97, in Main
2021-11-27T20:55:29.0252330Z     subprocess.check_call(compiler_driver_args, env=env)
2021-11-27T20:55:29.0253330Z   File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/subprocess.py", line 190, in check_call
2021-11-27T20:55:29.0254270Z     raise CalledProcessError(retcode, cmd)
```
I think there are some newer commits that update Chromium further which may fix this, as `qtwebengine` does build fine for Qt6 on Monterey.  Otherwise we'd have to figure out if there is a missing `LDFLAG`.

**Comment by Bo98** [comment]
Later Chromium commits delete `base::string16` entirely (for a different reason). That would be a somewhat large patch to backport.

I'm guessing this is linked to changes to `shrink_to_fit` in the macOS 12 SDK. I believe it's supposed to be inlined so it shouldn't be trying to externally link it here. I'm trying to slimline it down to a small reproducible case but I've not had much success in doing so.

**Comment by danielnachun** [comment]
I may be mistaken about this, but it seems to me that the way they update the Chromium version is by specifying a revision of the `qtwebengine-chromium` subproject: https://invent.kde.org/qt/qt/qtwebengine/-/commit/ee3252238c200d84461d5da71bb413b798c117f3.  If the revision they are using for the latest commit is new enough then maybe a backport isn't needed?

**Comment by danielnachun** [comment]
This is the newest version I think for `qtwebengine` 5.15: https://invent.kde.org/qt/qt/qtwebengine/-/commit/604f42c37b36a4674f953665f84872e4d83e0316.

**Comment by Bo98** [comment]
Qt 5.15 follow Chromium 87 with security patches backported. There's likely a compatibility reason to not update further than that. Chromium made `string16` changes in version 89.

The commit you've linked is included in 5.15.7 (fetched via git submodules).

**Comment by danielnachun** [comment]
Hmm, I wonder if we should reach out to KDE upstream about this.  I think they are supporting Qt 5 on all platforms for now until they have migrated fully over to Qt 6.

**Comment by Bo98** [comment]
The interesting piece of info would be whether this is reproducible with LLVM Clang. There shouldn't be really be much differences in terms of C++ headers between Apple and LLVM.

**Comment by Bo98** [comment]
This is now being tracked upstream: https://bugreports.qt.io/browse/QTBUG-98813

**Comment by danielnachun** [comment]
Interestingly this comment https://trac.macports.org/ticket/63725#comment:12 speculates that it could be a bug in `libc++` for the macOS 12 SDK.  I don't think we'd want to open the can of worms that would come with linking to brewed `libc++` from LLVM, but that comment also proposes a very simple fix that we could attempt to patch in.

**Comment by Bo98** [comment]
I do agree somethings wrong but I'm struggling to reproduce anything small that I can report to Apple. Tried extracting the string16 files and compiled them with something that called `shrink_to_fit()` and got no error. The C++ headers in the SDK also largely match LLVM's so in theory the issue should also affect LLVM 13's libc++, but I've not confirmed this yet.

The workaround does seem reasonable (`shrink_to_fit()` -> `reserve(0)`) though.

**Comment by danielnachun** [comment]
The other part of that comment mentioned that `shink_to_fit()` and `reserve(0)` are synonymous in C++17 and earlier, and then when the poster compiled ` src/3rdparty/chromium/base/strings/string16.cc` outside of the Chromium build system, he got an object that used `base::string16::reserve` instead of `base::string16::shrink_to_fit`.  

That makes me think that something about the way the Chromium build system calls `clang` or the linker is preventing that symbol substitution from happening.  I am suspicious of `-Wcrl,strip,-x,-S`, which I don't think is removed by the superenv and looks like it tries to strip unused symbols.  Perhaps `shrink_to_fit` is stripped under the assumption that it was converted to `reserve(0)` but for some reason that fails to happen and the object still contains `shrink_to_fit`?

**Comment by Bo98** [comment]
The correct behaviour is that `shrink_to_fit` is an internal symbol and `reserve(unsigned long)` is an external one (regardless of SDK). `shrink_to_fit` is inlined due for ABI compatibility reasons.

> I am suspicious of `-Wcrl,strip,-x,-S`, which I don't think is removed by the superenv and looks like it tries to strip unused symbols.

`qt@5` currently bypasses the superenv because it invokes `xcrun -find clang`. It's a bit tricky dealing with arg parsing in Bash, but I think we should try to finally make the `xcrun` shim return our superenv clang if `-find` is passed. I will look into making this change and see what breaks here (potentially many things but we should try).

**Comment by danielnachun** [comment]
I didn't realize `xcrun -find` bypassed the superenv.  That seems like it would be worth investigating so we can better control the flags being passed.

If it turns out that symbol stripping is the problem here, I'm wondering if we'll have to look into explicitly removing symbol stripping flags in the superenv.  To me whatever disk space is saved by stripping unused symbols is not worth the risk of obscure, complex failures when the symbol stripping is too aggressive.

**Comment by Bo98** [comment]
The build issue is fixed in 5.15.8, which will be released this month. I imagine next week to get it done before the holidays.

**Comment by johnnyutahh** [comment]
I'm trying to install `pyqt` and the install is currently breaking,  possibly due to this bug/issue. Can anyone suggest a workaround?

(I realize I'm posing a user-based question in a development-based discussion; please pardon me if my query is inappropriate. In any case, thanks all for the efforts to make this project go, it's truly appreciated.)

**Comment by jpfeuffer** [comment]
Are there any updates on Qt5 with WebEngine for Arm64? All the upstream bugs have been fixed. Can this PR be reopened and updated?

**Comment by Bo98** [comment]
Yes, I'll get back to this tonight/tomorrow. 5.15.8 has been released with the fixes.

**Comment by Bo98** [comment]
> Are there any updates on Qt5 with WebEngine for Arm64?

Just realised I read this wrong, this part isn't actually fully fixed yet: https://bugreports.qt.io/browse/QTBUG-97961. We may however be able to work around it - I'll have a look.

Monterey issues have however been fixed in 5.15.8.

**Comment by jpfeuffer** [comment]
Oh! I see! I was also confused by all the different upstream issues. Sorry for making noise.
I was basing my thoughts on this issue: https://bugreports.qt.io/browse/QTBUG-85537

**Comment by danielnachun** [comment]
One potential snag we may run into here now - as far as I know, `qtwebengine` for `qt@5` will still require Python 2 to build, and Python 2 will go away with the 12.3 update.  So if we want to get `qtwebengine` support for `qt@5` on Monterey, we'll have to do it before our CI migrates to 12.3.

**Comment by Bo98** [comment]
Merged: b0a871aadd4d51af8fff4c0580d1fafbaa72bd16

**Pull request closed without merging**